torch/_custom_op/autograd.py

```diff
@@ -0,0 +1,222 @@
+            return kernel(*args, **kwargs)
+        if custom_op._has_impl('save_for_backward') or custom_op._has_impl('backward'):
+            missing = (
+                'save_for_backward' if custom_op._has_impl('backward')
+                else 'backward'
+            )
+            found = 'save_for_backward' if missing == 'backward' else 'backward'
+            loc = custom_op._get_impl(found).location
+            raise RuntimeError(
+                f"We found a '{found}' registration for {custom_op} at "
+                f"{loc} but were unable to find a '{missing}' registration. "
+                f"To use the CustomOp API to register a backward formula, "
+                f"please provide us both a backward function and a "
+                f"'save for backward' function via `impl_backward` and "
+                f"`impl_save_for_backward` respectively.")
+        return autograd_not_implemented(*args, **kwargs)
+    return inner
+
+
+def construct_autograd_kernel(
+        schema,
+        output_differentiability,
+        forward_op,
+        save_for_backward_fn,
+        backward_fn):
+
+    def apply(*args):
+        flat_args, spec = pytree.tree_flatten(args)
+
+        def forward(ctx, *flat_args):
+            ctx.set_materialize_grads(True)
+            args = pytree.tree_unflatten(flat_args, spec)
+            guard = torch._C._AutoDispatchBelowAutograd()
+            try:
+                output = forward_op(*args)
+            finally:
+                del guard
+
+            # We use the info about args to give better error messages in backward
+            args_info = namedtuple_args(
+                schema, pytree.tree_map(lambda arg: type(arg), args))
+
+            save_for_backward_fn_inputs = namedtuple_args(schema, args)
+            to_save = save_for_backward_fn(save_for_backward_fn_inputs, output)
+
+            save_pytree_for_backward(ctx, (to_save, args_info))
+
+            # Output must be one or more Tensors, no TensorList (yet)
+            if output_differentiability is not None:
+                if isinstance(output, tuple):
+                    assert len(output_differentiability) == len(output)
+                    for differentiable, out in zip(output_differentiability, output):
+                        if not differentiable:
+                            ctx.mark_non_differentiable(out)
+                else:
+                    assert len(output_differentiability) == 1
+                    if not output_differentiability[0]:
+                        ctx.mark_non_differentiable(output)
+
+            return output
+
+        def backward(ctx, *grads):
+            saved, args_info = unpack_saved(ctx)
+            # There is nothing on the ctx object for now, it is just there so
+            # that we can add additional things in the future.
+            inner_ctx = object()
+            grad_inputs_dict = backward_fn(inner_ctx, saved, *grads)
```

Comment:
Curious what the API would look like with the implicit autograd API, are there any bc-breaking things we should be worried about when we transition to that one?

Reply:
By "implicit" do you mean "magic"?

We can just name the magic API something different; then there are no BC-breaking changes. But I might be misunderstanding the question
```py
@myop.impl_backward_unified(output_differentiability=...)
def backward(grad, x):
    return grad * x.cos()
```

Reply:
Oh yeah thats what I meant. That plan sounds good.